src/functional/ffront/foast_pretty_printer.py

```diff
@@ -0,0 +1,195 @@
+    # Shifts
+    #  not used
+    # Bitwise AND
+    #  not used
+    # Bitwise XOR
+    #  not used
+    # Bitwise OR
+    #  not used
+    # Comparisons, including membership tests and identity tests
+    (foast.Compare, "less"): 7,
+    (foast.Compare, "less_equal"): 7,
+    (foast.Compare, "greater"): 7,
+    (foast.Compare, "greater_equal"): 7,
+    (foast.Compare, "not_eq"): 7,
+    (foast.Compare, "eq"): 7,
+    # Boolean NOT
+    (foast.UnaryOp, "not_"): 6,
+    # Boolean AND
+    (foast.BinOp, "and_"): 5,
+    # Boolean OR
+    (foast.BinOp, "or_"): 4,
+    # Conditional expression
+    foast.TernaryExpr: 3,
+    # Lambda expression
+    #  not used
+    # Assignment expression
+    #  not used
+}
+
+RIGHT_GROUPING = {(foast.BinOp, "power"), foast.TernaryExpr}
+
+
+class Group(enum.IntEnum):
+    LEFT = 0
+    RIGHT = 1
+
+
+def _property_identifier(node: foast.LocatedNode) -> PropertyIdentifier:
+    if isinstance(node, foast.BinOp):
+        return (foast.BinOp, node.op)
+    elif isinstance(node, foast.UnaryOp):
+        return (foast.UnaryOp, node.op)
+    elif isinstance(node, foast.Compare):
+        return (foast.Compare, node.op)
+
+    return type(node)
+
+
+class PrettyPrinter(codegen.TemplatedGenerator):
+    Symbol = as_fmt("{id}")
+
+    Name = as_fmt("{id}")
+
+    Constant = as_fmt("{value}")
+
+    Subscript = as_fmt("{value}[{index}]")
+
+    Attribute = as_fmt("{value}.{attr}")
+
+    TupleExpr = as_fmt("({', '.join(elts)}{',' if len(elts)==1 else ''})")
+
+    UnaryOp = as_fmt("{op}{operand}")
+
+    def visit_BinOp(self, node: foast.BinOp, **kwargs) -> str:
+        left = self._parenthesize(node.left, node, Group.LEFT)
+        right = self._parenthesize(node.right, node, Group.RIGHT)
+        return f"{left} {node.op} {right}"
+
+    def visit_Compare(self, node: foast.Compare, **kwargs) -> str:
+        left = self._parenthesize(node.left, node, Group.LEFT)
+        right = self._parenthesize(node.right, node, Group.RIGHT)
+        return f"{left} {node.op} {right}"
+
+    def visit_TernaryExpr(self, node: foast.TernaryExpr, **kwargs) -> str:
+        cond = self.visit(node.condition)
+        true_expr = self._parenthesize(node.true_expr, node, Group.LEFT)
+        false_expr = self._parenthesize(node.false_expr, node, Group.RIGHT)
+        return f"{true_expr} if {cond} else {false_expr}"
+
+    Call = as_fmt("{func}({', '.join(args)})")  # TODO: kwargs
+
+    Assign = as_fmt("{target} = {value}")
+
+    Return = as_fmt("return {value}")
+
+    FunctionDefinition = as_mako(
+        textwrap.dedent(
+            """
+            def ${id}(${', '.join(params_annotated)})${return_type}:
+              ${'\\n  '.join(body)}
+            """
+        ).strip()
+    )
+
+    def visit_FunctionDefinition(self, node: foast.FunctionDefinition, **kwargs):
+        params = self.visit(node.params)
+        types = [
+            str(param.type) if not isinstance(param.type, common_types.DeferredSymbolType) else None
```

Comment:
I like the condition to evaluate to True:
```suggestion
            None if isinstance(param.type, common_types.DeferredSymbolType) else str(param.type)
```
Also, are you sure that param will always have a type?

Reply:
This is a stylistic question, but I would argue that the reader of the code should be presented with the common case first for conditionals, which is to stringify the type here. For if statements, contrary to conditionals, the reader is first presented with the condition so one can easily follow along and I always suggest avoiding unnecessary negations. Not so much for conditionals where one is first presented with a value only after which the condition follows. I personally dislike the choice of the conditional syntax in Python for exactly that reason.